File: src/mem/slicc/symbols/Func.cc

/*
 * Copyright (c) 1999-2008 Mark D. Hill and David A. Wood
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met: redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer;
 * redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution;
 * neither the name of the copyright holders nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Func.cc
 *
 * Description: See Func.hh
 *
 * $Id$
 *
 */

#include "mem/slicc/symbols/Func.hh"
#include "mem/slicc/symbols/SymbolTable.hh"
#include "mem/slicc/generator/fileio.hh"
#include "mem/slicc/symbols/StateMachine.hh"

Func::Func(string id, const Location& location,
           Type* type_ptr, const Vector<Type*>& param_type_vec,
           const Vector<string>& param_string_vec, string body,
           const Map<string, string>& pairs, StateMachine* machine_ptr)
  : Symbol(id, location, pairs)
{
  m_type_ptr = type_ptr;
  m_param_type_vec = param_type_vec;
  m_param_string_vec = param_string_vec;
  m_body = body;
  m_isInternalMachineFunc = false;

  if (machine_ptr == NULL) {
    m_c_ident = id;
  } else if (existPair("external") || existPair("primitive")) {
    m_c_ident = id;
  } else {
    m_machineStr = machine_ptr->toString();
    m_c_ident = m_machineStr + "_" + id;  // Append with machine name
    m_isInternalMachineFunc = true;
  }
}

void Func::funcPrototype(string& code) const
{
  if (isExternal()) {
    // Do nothing
  } else {
    string return_type = m_type_ptr->cIdent();
    Type* void_type_ptr = g_sym_table.getType("void");
    if (existPair("return_by_ref") && (m_type_ptr != void_type_ptr)) {
      return_type += "&";
    }
    code += return_type + " " + cIdent() + "(";
    int size = m_param_string_vec.size();
    for(int i=0; i<size; i++) {
      // Generate code
      if (i != 0) {
        code += ", ";
      }
      code += m_param_string_vec[i];
    }
    code += ");\n";
  }
}

// This write a function of object Chip
void Func::writeCFiles(string path) const
{
  if (isExternal()) {
    // Do nothing
  } else {
    ostringstream out;

    // Header
    out << "/** Auto generated C++ code started by "<<__FILE__<<":"<<__LINE__<< " */" << endl;
    out << endl;
    out << "#include \"mem/protocol/Types.hh\"" << endl;
    out << "#include \"mem/protocol/Chip.hh\"" << endl;
    if (m_isInternalMachineFunc) {
      out << "#include \"" << m_machineStr << "_Controller.hh\"" << endl;
    }
    out << endl;

    // Generate function header
    string code;
    Type* void_type_ptr = g_sym_table.getType("void");
    string return_type = m_type_ptr->cIdent();
    code += return_type;
    if (existPair("return_by_ref") && m_type_ptr != void_type_ptr) {
      code += "&";
    }
    if (!m_isInternalMachineFunc) {
      code += " Chip::" + cIdent() + "(";
    } else {
      code += " " + m_machineStr + "_Controller::" + cIdent() + "(";
    }
    int size = m_param_type_vec.size();
    for(int i=0; i<size; i++) {
      // Generate code
      if (i != 0) {
        code += ", ";
      }
      code += m_param_string_vec[i];
    }
    code += ")";

    // Function body
    code += "\n{\n";
    code += m_body;
    code += "}\n";
    out << code << endl;

    // Write it out
    conditionally_write_file(path + cIdent() + ".cc", out);
  }
}

void Func::print(ostream& out) const
{
}

File: src/mem/slicc/symbols/Func.py
# Copyright (c) 1999-2008 Mark D. Hill and David A. Wood
# Copyright (c) 2009 The Hewlett-Packard Development Company
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met: redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer;
# redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution;
# neither the name of the copyright holders nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from slicc.symbols.Symbol import Symbol
from slicc.symbols.Type import Type

class Func(Symbol):
    def __init__(self, table, ident, name, location, return_type, param_types,
                 param_strings, body, pairs):
        super(Func, self).__init__(table, ident, location, pairs)
        self.return_type = return_type
        self.param_types = param_types
        self.param_strings = param_strings
        self.body = body
        self.isInternalMachineFunc = False
        self.c_ident = ident
        self.c_name = name
        self.class_name = ""

    def __repr__(self):
        return ""

    @property
    def prototype(self):
        if "external" in self:
            return ""

        return_type = self.return_type.c_ident
        void_type = self.symtab.find("void", Type)
        if "return_by_ref" in self and self.return_type != void_type:
            return_type += "&"
        elif "return_by_pointer" in self and self.return_type != void_type:
            return_type += "*"

        return "%s %s(%s);" % (return_type, self.c_name,
                               ", ".join(self.param_strings))

    def writeCodeFiles(self, path, includes):
        return

    def checkArguments(self, args):
        if len(args) != len(self.param_types):
            self.error("Wrong number of arguments passed to function : '%s'" +\
                       " Expected %d, got %d", self.c_ident,
                       len(self.param_types), len(args))

        cvec = []
        type_vec = []
        for expr,expected_type in zip(args, self.param_types):
            # Check the types of the parameter
            actual_type,param_code = expr.inline(True)
            if str(actual_type) != 'OOD' and \
               str(actual_type) != str(expected_type) and \
               str(actual_type["interface"]) != str(expected_type):
                expr.error("Type mismatch: expected: %s actual: %s" % \
                           (expected_type, actual_type))
            cvec.append(param_code)
            type_vec.append(expected_type)

        return cvec, type_vec

    def generateCode(self):
        '''This write a function of object Chip'''
        if "external" in self:
            return ""

        code = self.symtab.codeFormatter()

        # Generate function header
        void_type = self.symtab.find("void", Type)
        return_type = self.return_type.c_ident
        if "return_by_ref" in self and self.return_type != void_type:
            return_type += "&"
        if "return_by_pointer" in self and self.return_type != void_type:
            return_type += "*"

        params = ', '.join(self.param_strings)

        code('''
$return_type
${{self.class_name}}::${{self.c_name}}($params)
{
${{self.body}}
}
''')
        return str(code)

__all__ = [ "Func" ]
